linkerd/proxy/identity/src/metrics.rs

```diff
@@ -0,0 +1,53 @@
+use crate::CrtKey;
+use linkerd2_metrics::{metrics, FmtMetrics, Gauge};
+use std::{fmt, time::UNIX_EPOCH};
+use tokio::sync::watch;
+
+#[derive(Debug, Clone)]
+pub struct Report {
+    crt_key_watch: Option<watch::Receiver<Option<CrtKey>>>,
+}
+
+metrics! {
+    identity_cert_expiration_timestamp_seconds: Gauge {
+        "Time when the this proxy's current mTLS identity certificate will expire (in seconds since the UNIX epoch)"
+    }
+}
+
+impl Report {
+    pub(crate) fn new(watch: watch::Receiver<Option<CrtKey>>) -> Self {
+        Self {
+            crt_key_watch: Some(watch),
+        }
+    }
+
+    pub fn disabled() -> Self {
+        Self {
+            crt_key_watch: None,
+        }
+    }
+}
+
+impl FmtMetrics for Report {
+    fn fmt_metrics(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let dur = if let Some(watch) = self.crt_key_watch.as_ref() {
+            if let Some(ref crt_key) = *(watch.borrow()) {
+                crt_key
+                .expiry()
+                .duration_since(UNIX_EPOCH)
+                .map_err(|error| {
+                    tracing::warn!(%error, "an identity would expire before the beginning of the UNIX epoch, something is probably wrong");
+                    fmt::Error
+                })?
+            } else {
+                return Ok(());
+            }
+        } else {
+            return Ok(());
+        };
```

Comment:
tioli: returning error here will also prevent other reports from being formatted. Would it be better to panic? Or perhaps we should factor this all into a `fn` that returns `Option<Duration>` and then make the format body `if let Some(d) = self.expiry() { ... }`

Reply:
> tioli: returning error here will also prevent other reports from being formatted.

we're returning `Ok` here, so it should continue formatting any other reports?

> Would it be better to panic? Or perhaps we should factor this all into a `fn` that returns `Option<Duration>` and then make the format body `if let Some(d) = self.expiry() { ... }`

yeah, i wasn't sure whether it was better to panic or not. On one hand, this definitely shouldn't ever actually happen, but on the other hand, I dunno if it's a good idea to bring down the whole proxy if one metric can't be formatted...it's not _technically_ a critical error at all. on the *third* hand, if we ever see an expiration timestamp that's before the Unix epoch, something else is probably going wrong...

Reply:
> we're returning Ok here, so it should continue formatting any other reports?

error returned here https://github.com/linkerd/linkerd2-proxy/pull/787/files/b324a9b1693c3af86c9abcf919b277b81d7bd25a#diff-dbf40c2f9158dbeb59c53424a494292110865ee6d84bb197ab3f515afb644812R41